File: Standalone tools/investment-mobile/native/oms/Oms.cs
// 星澄 AI 投資管理與自動操盤系統 — order management state machine (C#).
//
// Decision-free orchestration: the same states and transitions as the
// Python OMS (src/.../trading/oms.py). Trading orchestration, order
// management and broker connections are the C# responsibility per the
// language division of labour.
//
// Fixed pipeline: TradeProposal -> RiskDecision -> OrderRequest ->
// (mode gate) -> BrokerAdapter (Live) / simulated account (Paper) ->
// OrderReceipt -> Execution -> Portfolio. Shadow records the decided
// order but never submits; Analysis stops after the risk decision.

namespace InvestmentMobile.Oms;

public enum TradingMode { Analysis, Shadow, Paper, Live }

public enum OrderStatus
{
    Created,
    RiskRejected,
    ModeBlocked,
    AdapterDenied,
    Submitted,
    PartialFilled,
    Filled,
    Cancelled,
}

/// <summary>Formal trade proposal — the only AI-reachable shape.</summary>
public sealed record TradeProposal(
    string InstrumentId,
    string Market,
    string Side,
    double Quantity,
    double? Price,
    string StrategyId,
    string SignalId,
    string AccountId = "")
{
    public double EffectiveNotional =>
        Quantity * (Price ?? 0.0);
}

public sealed record RiskDecision(bool Approved, IReadOnlyList<string> Reasons);

public interface IRiskGate
{
    RiskDecision Evaluate(TradeProposal proposal);
}

public interface IBrokerAdapter
{
    string BrokerId { get; }
    string Market { get; }
    bool ApiVerified { get; }
    BrokerOrderResult PlaceOrder(OrderRequest order);
}

public sealed record BrokerOrderResult(bool Ok, string? ErrorCode, string? BrokerOrderId);

public sealed class OrderRequest
{
    public required TradeProposal Proposal { get; init; }
    public string OrderId { get; } = $"ord-{Guid.NewGuid():N}"[..16];
    public OrderStatus Status { get; set; } = OrderStatus.Created;
    public string BrokerOrderId { get; set; } = "";
    public string Rejection { get; set; } = "";
}

public sealed record OrderReceipt(
    string OrderId,
    string BrokerOrderId,
    bool Simulated);

public sealed record Execution(
    string OrderId,
    string InstrumentId,
    string Market,
    string Side,
    double Quantity,
    double Price,
    string AccountId,
    bool Simulated);

/// <summary>
/// Order pipeline: risk gate -> mode gate -> simulated fill (Paper only)
/// or verified broker dispatch (Live). Analysis records the decision
/// evidence; Shadow records the order but never submits.
/// </summary>
public sealed class OrderManagementSystem
{
    private readonly IRiskGate _risk;
    private readonly IReadOnlyDictionary<string, IBrokerAdapter> _adapters;
    private readonly List<OrderRequest> _orders = new();
    private readonly List<Execution> _executions = new();

    public TradingMode Mode { get; set; } = TradingMode.Analysis;

    public OrderManagementSystem(
        IRiskGate risk, IReadOnlyDictionary<string, IBrokerAdapter> adapters)
    {
        _risk = risk;
        _adapters = adapters;
    }

    public IReadOnlyList<OrderRequest> Orders => _orders;
    public IReadOnlyList<Execution> Executions => _executions;

    public OrderRequest Submit(TradeProposal proposal)
    {
        var order = new OrderRequest { Proposal = proposal };
        _orders.Add(order);

        var decision = _risk.Evaluate(proposal);
        if (!decision.Approved)
        {
            order.Status = OrderStatus.RiskRejected;
            order.Rejection = string.Join("; ", decision.Reasons);
            return order;
        }

        if (Mode == TradingMode.Analysis)
        {
            order.Status = OrderStatus.ModeBlocked;
            order.Rejection = "mode ANALYSIS blocks order submission";
            return order;
        }

        // Shadow: the decided order is recorded but never submitted.
        if (Mode == TradingMode.Shadow)
        {
            order.Status = OrderStatus.ModeBlocked;
            order.Rejection = "mode SHADOW records decisions without submission";
            return order;
        }

        if (Mode == TradingMode.Paper)
        {
            // Dedicated simulated account — never touches a broker.
            order.Status = OrderStatus.Filled;
            _executions.Add(new Execution(
                order.OrderId, proposal.InstrumentId, proposal.Market,
                proposal.Side, proposal.Quantity, proposal.Price ?? 0.0,
                $"paper-{proposal.Market}", Simulated: true));
            return order;
        }

        // Live: dispatch only through an API-verified adapter.
        if (!_adapters.TryGetValue(proposal.Market, out var adapter))
        {
            order.Status = OrderStatus.AdapterDenied;
            order.Rejection = $"no broker adapter for market '{proposal.Market}'";
            return order;
        }

        var result = adapter.PlaceOrder(order);
        if (!result.Ok)
        {
            order.Status = OrderStatus.AdapterDenied;
            order.Rejection = result.ErrorCode ?? "ADAPTER_DENIED";
            return order;
        }

        order.Status = OrderStatus.Submitted;
        order.BrokerOrderId = result.BrokerOrderId ?? "";
        return order;
    }
}

File: Standalone tools/investment-mobile/src/backend/services/investment_mobile/trading/oms.py
"""Order Management System — decision-free order state machine.

Mirrors the C# orchestration component (``native/oms``): the same states
and transitions. Every transition is journaled to the trading audit and,
when bound, forwarded to ai-assistant as the authoritative record.

Pipeline: intent → risk evaluation → mode gate → adapter dispatch (LIVE,
verified only) or simulated fill (PAPER/SHADOW).
"""

from __future__ import annotations

import time
from typing import Any

from .audit import TradingAudit
from .broker.base import BrokerRegistry
from .contracts import Fill, Order, OrderIntent, OrderStatus
from .modes import ModeGate
from .portfolio_engine import PortfolioEngine
from .risk_engine import RiskEngine


class OrderManagementSystem:
    def __init__(
        self,
        *,
        mode_gate: ModeGate,
        risk_engine: RiskEngine,
        portfolio: PortfolioEngine,
        brokers: BrokerRegistry,
        audit: TradingAudit,
    ) -> None:
        self._mode_gate = mode_gate
        self._risk = risk_engine
        self._portfolio = portfolio
        self._brokers = brokers
        self._audit = audit
        self._orders: dict[str, Order] = {}
        self._fills: list[Fill] = []
        self._daily_order_count = 0
        self._daily_realized_pnl = 0.0
        self._day = time.strftime("%Y-%m-%d")

    # ------------------------------------------------------------------
    def _rollover(self) -> None:
        today = time.strftime("%Y-%m-%d")
        if today != self._day:
            self._day = today
            self._daily_order_count = 0
            self._daily_realized_pnl = 0.0

    def submit(self, intent: OrderIntent) -> dict[str, Any]:
        self._rollover()
        order = Order(intent=intent)
        self._orders[order.order_id] = order
        self._audit.record("order.created", order.to_dict())

        # 1. risk gate — always evaluated, in every mode.
        decision = self._risk.evaluate(
            intent,
            self._portfolio.positions(),
            daily_order_count=self._daily_order_count,
            daily_realized_pnl=self._daily_realized_pnl,
        )
        self._audit.record(
            "risk.evaluated",
            {"order_id": order.order_id, **decision.to_dict()},
        )
        if not decision.approved:
            order.status = OrderStatus.RISK_REJECTED.value
            order.rejection = "; ".join(decision.reasons)
            self._audit.record("order.rejected", order.to_dict())
            return {"ok": False, "order": order.to_dict(), "decision": decision.to_dict()}

        # 2. mode gate — ANALYSIS never submits.
        if not self._mode_gate.allows("orders"):
            order.status = OrderStatus.MODE_BLOCKED.value
            order.rejection = f"mode {self._mode_gate.mode.value} blocks order submission"
            self._audit.record("order.mode_blocked", order.to_dict())
            return {"ok": False, "order": order.to_dict(), "decision": decision.to_dict()}

        self._daily_order_count += 1

        # 3a. SHADOW — record the would-be decision, no fill.
        if self._mode_gate.mode.value == "SHADOW":
            order.status = OrderStatus.SUBMITTED.value
            order.rejection = ""
            self._audit.record("order.shadow", order.to_dict())
            return {"ok": True, "order": order.to_dict(), "decision": decision.to_dict()}

        # 3b. PAPER — simulated fill at the intent price.
        if self._mode_gate.mode.value == "PAPER":
            fill = Fill(
                order_id=order.order_id,
                instrument=intent.instrument,
                market=intent.market,
                side=intent.side,
                quantity=intent.quantity,
                price=float(intent.price or 0.0),
                simulated=True,
            )
            order.status = OrderStatus.FILLED.value
            self._fills.append(fill)
            self._portfolio.apply_fill(fill)
            self._audit.record("order.filled", {"order": order.to_dict(), "fill": fill.to_dict()})
            return {
                "ok": True,
                "order": order.to_dict(),
                "fill": fill.to_dict(),
                "decision": decision.to_dict(),
            }

        # 3c. LIVE — dispatch through the verified adapter only.
        adapter = self._brokers.for_market(intent.market)
        if adapter is None:
            order.status = OrderStatus.ADAPTER_DENIED.value
            order.rejection = f"no broker adapter for market '{intent.market}'"
            self._audit.record("order.adapter_denied", order.to_dict())
            return {"ok": False, "order": order.to_dict(), "decision": decision.to_dict()}
        result = adapter.place_order(order, intent)
        if not result.get("ok"):
            order.status = OrderStatus.ADAPTER_DENIED.value
            order.rejection = str(result.get("error_code") or "ADAPTER_DENIED")
            self._audit.record("order.adapter_denied", {**order.to_dict(), "adapter": result})
            return {"ok": False, "order": order.to_dict(), "decision": decision.to_dict(), "adapter": result}
        order.status = OrderStatus.SUBMITTED.value
        order.broker_order_id = str(result.get("broker_order_id") or "")
        self._audit.record("order.submitted", {**order.to_dict(), "adapter": result})
        return {"ok": True, "order": order.to_dict(), "decision": decision.to_dict(), "adapter": result}

    # ------------------------------------------------------------------
    def cancel(self, order_id: str) -> dict[str, Any]:
        order = self._orders.get(str(order_id))
        if order is None:
            return {"ok": False, "error_code": "ORDER_NOT_FOUND"}
        if order.status in (
            OrderStatus.FILLED.value,
            OrderStatus.CANCELLED.value,
            OrderStatus.RISK_REJECTED.value,
        ):
            return {"ok": False, "error_code": "ORDER_TERMINAL", "order": order.to_dict()}
        order.status = OrderStatus.CANCELLED.value
        order.updated_at = time.time()
        self._audit.record("order.cancelled", order.to_dict())
        return {"ok": True, "order": order.to_dict()}

    def orders(self, limit: int = 100) -> list[dict[str, Any]]:
        orders = sorted(self._orders.values(), key=lambda o: o.created_at, reverse=True)
        return [o.to_dict() for o in orders[: max(1, int(limit))]]

    def fills(self) -> list[Fill]:
        return list(self._fills)
